dna/models.py

```diff
@@ -394,6 +394,205 @@ def predict_rank(self, data, *, batch_size, verbose):
+
+        metafeatures_length = len(train_data[0]['metafeatures'])
+        model = DAGRNN(rnn_input_size=num_primitives, hidden_state_size=metafeatures_length, output_layer_size=1,
+                       n_layers=2, dropout=0.1, bidirectional=True)
+        model.cuda()
+        return model
+
+    def _get_primitive_name_to_enc(self, train_data):
+        primitive_names = set()
+
+        # Get a set of all the primitives in the train set
+        for instance in train_data:
+            primitives = instance[self.pipeline_key][self.steps_key]
+            for primitive in primitives:
+                primitive_name = primitive[self.prim_name_key]
+                primitive_names.add(primitive_name)
+
+        # Get one hot encodings of all the primitives
+        num_primitives = len(primitive_names)
+        encoding = np.zeros((num_primitives, num_primitives))
+        for i in range(num_primitives):
+            encoding[i][i] = 1
+
+        # Create a mapping of primitive names to one hot encodings
+        self.primitive_name_to_enc = {}
+        for (primitive_name, primitive_encoding) in zip(primitive_names, encoding):
+            self.primitive_name_to_enc[primitive_name] = primitive_encoding
+
+        return num_primitives
+
+    def _get_optimizer(self, learning_rate):
+        return torch.optim.Adam(self._model.parameters(), lr=learning_rate)
+
+    def _get_data_loader(self, data, batch_size, drop_last):
+        dataset_params = {
+            "features_key": "metafeatures",
+            "target_key": self.target_key,
+            "task_type": self._task_type,
+            "device": self.device,
+            'prim_to_enc': self.primitive_name_to_enc
+        }
+
+        return RNNDataLoader(data=data,
+                             group_key=self.batch_group_key,
+                             dataset_params=dataset_params,
+                             batch_size=batch_size,
+                             drop_last=drop_last,
+                             shuffle=True,
+                             seed=self.seed)
+
+    def predict_regression(self, data, *, batch_size, verbose):
+        if self._model is None:
+            raise Exception('model not fit')
+
+        data_loader = self._get_data_loader(data, batch_size, False)
+        predictions, targets = self._predict_epoch(data_loader, self._model, verbose=verbose)
+
+        return predictions
+
+    def predict_rank(self, data, *, batch_size, verbose):
+        if self._model is None:
+            raise Exception('model not fit')
+
+        data_loader = self._get_data_loader(data, batch_size, False)
+        predictions, targets = self._predict_epoch(data_loader, self._model, verbose=verbose)
+        ranks = utils.rank(np.array(predictions))
+        return {
+            'pipeline_id': [instance['pipeline']['id'] for instance in data],
+            'rank': ranks,
+        }
+
+
+class DAGRNN(nn.Module):
```

Comment:
If you are going to add the DAGRNN in this pull request, also add its usage. So add it to the `get_model` method and the test.sh file. If you don't think DAGRNN is ready for that, remove it from this pull request.

Reply:
If I understand right, the DAGRNN is equivalent to the K-nearest datasets model for the AutoSklearn, or the DNA model that underlies the DNA_Regression model.  The RNN_Regression model uses this model for its forward functions, etc.  It is not a stand alone model and shouldn't be placed in the test.sh file.  A doc string describing its purpose would be useful though.

Reply:
Oh, I see. My bad. I thought the RNNRegressionModel was for straight pipelines and the DAGRNN was for all pipelines, but now I see my mistake. While not strictly necessary, organizing the code so that the DAGRNN class code comes before the RNNRegressionModel code in this file would make that clearer.

Docstrings are always good.